File: ios/Sparks/BenchmarkEngine.swift
import Foundation
import UIKit

// Default durations (seconds). Must be kept in sync with the Android
// implementation in android/app/src/main/cpp/benchmark_engine.* so that
// results are comparable across platforms.
enum BenchmarkTiming {
    static let warmupSec: Double = 3.0
    static let measureSec: Double = 10.0
    static let cooldownSec: Double = 2.0
}

struct ShaderBenchResult: Codable {
    let index: Int
    let name: String
    let avgFps: Double
    let onePctLowFps: Double
    let medianFrameMs: Double
    let p99FrameMs: Double
    let frames: Int
    let droppedFrames: Int
    let skipped: Bool
}

struct BenchmarkReport: Codable {
    struct DeviceInfo: Codable {
        let os: String
        let model: String
        let gpu: String
    }
    struct BenchConfig: Codable {
        let resolution: [Int]
        let halfRes: Bool
        let vsync: Bool
        let warmupSec: Double
        let measureSec: Double
        let cooldownSec: Double
    }
    let version: Int
    let timestamp: String
    let device: DeviceInfo
    let config: BenchConfig
    let thermalStateStart: String
    let thermalStateEnd: String
    let shaders: [ShaderBenchResult]
    let overallScore: Double
}

enum BenchmarkMode {
    case singleShader(Int)
    case allShaders
}

enum BenchmarkPhase {
    case idle, warmup, measure, cooldown, done
}

final class BenchmarkEngine {

    private(set) var phase: BenchmarkPhase = .idle
    private(set) var results: [ShaderBenchResult] = []

    private var shaderNames: [String] = []
    private var shaderQueue: [Int] = []
    private var queuePosition: Int = 0
    private var phaseStartTime: CFAbsoluteTime = 0
    private var lastFrameTime: CFAbsoluteTime = 0
    private var frameTimes: [Double] = []
    private var thermalStart: ProcessInfo.ThermalState = .nominal
    private var startedAt: Date = Date()

    var isRunning: Bool { phase != .idle && phase != .done }
    var isDone: Bool { phase == .done }

    var totalShadersInRun: Int { shaderQueue.count }
    var currentPositionOneBased: Int { min(queuePosition + 1, shaderQueue.count) }

    var activeShaderIndex: Int? {
        guard isRunning, queuePosition < shaderQueue.count else { return nil }
        return shaderQueue[queuePosition]
    }

    func start(mode: BenchmarkMode,
               shaderNames: [String],
               pipelineAvailable: (Int) -> Bool) {
        self.shaderNames = shaderNames
        self.results = []
        self.thermalStart = ProcessInfo.processInfo.thermalState
        self.startedAt = Date()

        var queue: [Int] = []
        switch mode {
        case .singleShader(let idx):
            if idx >= 0 && idx < shaderNames.count && pipelineAvailable(idx) {
                queue = [idx]
            }
        case .allShaders:
            queue = (0..<shaderNames.count).filter { pipelineAvailable($0) }
        }
        shaderQueue = queue
        queuePosition = 0
        frameTimes = []

        if queue.isEmpty {
            phase = .done
            return
        }
        phase = .warmup
        phaseStartTime = CFAbsoluteTimeGetCurrent()
    }

    func abort() {
        phase = .idle
        results = []
        shaderQueue = []
        queuePosition = 0
        frameTimes = []
    }

    /// Records frame time (present-to-present) during measure phase.
    /// Call at the start of each draw, before advancePhase.
    func recordPresentTime(now: CFAbsoluteTime) {
        guard phase == .measure else { return }
        let dt = now - lastFrameTime
        lastFrameTime = now
        // Sanity filter: drop absurd gaps (paused / backgrounded).
        if dt > 0 && dt < 1.0 {
            frameTimes.append(dt)
        }
    }

    /// Advances the state machine based on wall-clock elapsed time.
    /// Call each frame after recordPresentTime.
    func advancePhase(now: CFAbsoluteTime) {
        guard isRunning else { return }
        let elapsed = now - phaseStartTime
        switch phase {
        case .warmup:
            if elapsed >= BenchmarkTiming.warmupSec {
                phase = .measure
                phaseStartTime = now
                lastFrameTime = now
                frameTimes = []
            }
        case .measure:
            if elapsed >= BenchmarkTiming.measureSec {
                let idx = shaderQueue[queuePosition]
                results.append(makeResult(index: idx, frameTimes: frameTimes, skipped: false))
                phase = .cooldown
                phaseStartTime = now
            }
        case .cooldown:
            if elapsed >= BenchmarkTiming.cooldownSec {
                queuePosition += 1
                if queuePosition >= shaderQueue.count {
                    phase = .done
                } else {
                    phase = .warmup
                    phaseStartTime = now
                    frameTimes = []
                }
            }
        default:
            break
        }
    }

    var statusText: String {
        guard isRunning else { return "" }
        let total = shaderQueue.count
        let pos = currentPositionOneBased
        let name = currentShaderName()
        let phaseName: String
        let remaining: Double
        let elapsed = CFAbsoluteTimeGetCurrent() - phaseStartTime
        switch phase {
        case .warmup:
            phaseName = "warmup"
            remaining = max(0, BenchmarkTiming.warmupSec - elapsed)
        case .measure:
            phaseName = "measure"
            remaining = max(0, BenchmarkTiming.measureSec - elapsed)
        case .cooldown:
            phaseName = "cooldown"
            remaining = max(0, BenchmarkTiming.cooldownSec - elapsed)
        default:
            phaseName = ""
            remaining = 0
        }
        return String(format: "Bench %d/%d  %@  %@ %.1fs", pos, total, name, phaseName, remaining)
    }

    func currentShaderName() -> String {
        guard queuePosition < shaderQueue.count else { return "" }
        let idx = shaderQueue[queuePosition]
        return idx < shaderNames.count ? shaderNames[idx] : "?"
    }

    private func makeResult(index: Int, frameTimes: [Double], skipped: Bool) -> ShaderBenchResult {
        let name = index < shaderNames.count ? shaderNames[index] : "?"
        let count = frameTimes.count
        if skipped || count == 0 {
            return ShaderBenchResult(index: index, name: name,
                                     avgFps: 0, onePctLowFps: 0,
                                     medianFrameMs: 0, p99FrameMs: 0,
                                     frames: 0, droppedFrames: 0, skipped: skipped)
        }
        let totalTime = frameTimes.reduce(0, +)
        let avgFps = Double(count) / totalTime
        let sorted = frameTimes.sorted()
        let median = sorted[count / 2]
        let p99Idx = min(count - 1, Int((Double(count) * 0.99).rounded(.down)))
        let p99 = sorted[p99Idx]
        // 1% low FPS: mean of the slowest 1% of frames, inverted.
        let lowCount = max(1, count / 100)
        let lowSlice = sorted.suffix(lowCount)
        let lowAvg = lowSlice.reduce(0, +) / Double(lowCount)
        let onePctLowFps = lowAvg > 0 ? 1.0 / lowAvg : 0
        let droppedThreshold = median * 2.0
        let dropped = frameTimes.reduce(0) { $0 + ($1 > droppedThreshold ? 1 : 0) }
        return ShaderBenchResult(index: index, name: name,
                                 avgFps: avgFps,
                                 onePctLowFps: onePctLowFps,
                                 medianFrameMs: median * 1000.0,
                                 p99FrameMs: p99 * 1000.0,
                                 frames: count,
                                 droppedFrames: dropped,
                                 skipped: false)
    }

    /// Harmonic mean of avgFps × 100 across valid (non-skipped) results.
    func overallScore() -> Double {
        let valid = results.filter { !$0.skipped && $0.avgFps > 0 }
        if valid.isEmpty { return 0 }
        let sumReciprocal = valid.reduce(0.0) { $0 + 1.0 / $1.avgFps }
        return Double(valid.count) / sumReciprocal * 100.0
    }

    func makeReport(resolution: (Int, Int), halfRes: Bool, vsync: Bool, gpuName: String) -> BenchmarkReport {
        let fmt = ISO8601DateFormatter()
        let device = UIDevice.current
        return BenchmarkReport(
            version: 1,
            timestamp: fmt.string(from: startedAt),
            device: .init(os: "\(device.systemName) \(device.systemVersion)",
                          model: Self.modelIdentifier(),
                          gpu: gpuName),
            config: .init(resolution: [resolution.0, resolution.1],
                          halfRes: halfRes,
                          vsync: vsync,
                          warmupSec: BenchmarkTiming.warmupSec,
                          measureSec: BenchmarkTiming.measureSec,
                          cooldownSec: BenchmarkTiming.cooldownSec),
            thermalStateStart: Self.thermalString(thermalStart),
            thermalStateEnd: Self.thermalString(ProcessInfo.processInfo.thermalState),
            shaders: results,
            overallScore: overallScore()
        )
    }

    private static func thermalString(_ s: ProcessInfo.ThermalState) -> String {
        switch s {
        case .nominal: return "nominal"
        case .fair: return "fair"
        case .serious: return "serious"
        case .critical: return "critical"
        @unknown default: return "unknown"
        }
    }

    private static func modelIdentifier() -> String {
        var sysinfo = utsname()
        uname(&sysinfo)
        let mirror = Mirror(reflecting: sysinfo.machine)
        var id = ""
        for child in mirror.children {
            if let v = child.value as? Int8, v != 0 {
                id.append(Character(UnicodeScalar(UInt8(v))))
            }
        }
        return id
    }
}

File: android/app/src/main/cpp/benchmark_engine.cpp
#include "benchmark_engine.h"

#include <algorithm>
#include <cstdio>
#include <sstream>

namespace bench {

static std::string escapeJson(const std::string& s) {
    std::string out;
    out.reserve(s.size() + 2);
    for (char c : s) {
        switch (c) {
            case '"': out += "\\\""; break;
            case '\\': out += "\\\\"; break;
            case '\n': out += "\\n"; break;
            case '\r': out += "\\r"; break;
            case '\t': out += "\\t"; break;
            default:
                if (static_cast<unsigned char>(c) < 0x20) {
                    char buf[8];
                    std::snprintf(buf, sizeof(buf), "\\u%04x", c);
                    out += buf;
                } else {
                    out += c;
                }
        }
    }
    return out;
}

void BenchmarkEngine::start(Mode mode,
                            const std::vector<std::string>& shaderNames,
                            const PipelineAvailable& pipelineAvailable) {
    std::lock_guard<std::mutex> lock(mMutex);
    mNames = shaderNames;
    mResults.clear();
    mFrameTimes.clear();
    mQueue.clear();
    mQueuePos = 0;

    if (mode.kind == Mode::Single) {
        int idx = mode.singleIndex;
        if (idx >= 0 && idx < (int)shaderNames.size() && pipelineAvailable(idx)) {
            mQueue.push_back(idx);
        }
    } else {
        for (int i = 0; i < (int)shaderNames.size(); ++i) {
            if (pipelineAvailable(i)) mQueue.push_back(i);
        }
    }

    if (mQueue.empty()) {
        mPhase = Phase::Done;
        return;
    }
    mPhase = Phase::Warmup;
    mPhaseStart = Clock::now();
}

void BenchmarkEngine::abort() {
    std::lock_guard<std::mutex> lock(mMutex);
    mPhase = Phase::Idle;
    mResults.clear();
    mQueue.clear();
    mQueuePos = 0;
    mFrameTimes.clear();
}

bool BenchmarkEngine::isRunning() const {
    std::lock_guard<std::mutex> lock(mMutex);
    return mPhase != Phase::Idle && mPhase != Phase::Done;
}

bool BenchmarkEngine::isDone() const {
    std::lock_guard<std::mutex> lock(mMutex);
    return mPhase == Phase::Done;
}

void BenchmarkEngine::recordPresent(TimePoint now) {
    std::lock_guard<std::mutex> lock(mMutex);
    if (mPhase != Phase::Measure) return;
    double dt = std::chrono::duration<double>(now - mLastFrame).count();
    mLastFrame = now;
    if (dt > 0.0 && dt < 1.0) {
        mFrameTimes.push_back(dt);
    }
}

void BenchmarkEngine::advancePhase(TimePoint now) {
    std::lock_guard<std::mutex> lock(mMutex);
    if (mPhase == Phase::Idle || mPhase == Phase::Done) return;
    double elapsed = std::chrono::duration<double>(now - mPhaseStart).count();
    switch (mPhase) {
        case Phase::Warmup:
            if (elapsed >= kWarmupSec) {
                mPhase = Phase::Measure;
                mPhaseStart = now;
                mLastFrame = now;
                mFrameTimes.clear();
            }
            break;
        case Phase::Measure:
            if (elapsed >= kMeasureSec) {
                int idx = mQueue[mQueuePos];
                mResults.push_back(makeResultLocked(idx, mFrameTimes, false));
                mPhase = Phase::Cooldown;
                mPhaseStart = now;
            }
            break;
        case Phase::Cooldown:
            if (elapsed >= kCooldownSec) {
                mQueuePos++;
                if (mQueuePos >= (int)mQueue.size()) {
                    mPhase = Phase::Done;
                } else {
                    mPhase = Phase::Warmup;
                    mPhaseStart = now;
                    mFrameTimes.clear();
                }
            }
            break;
        default: break;
    }
}

int BenchmarkEngine::activeShaderIndex() const {
    std::lock_guard<std::mutex> lock(mMutex);
    if (mPhase == Phase::Idle || mPhase == Phase::Done) return -1;
    if (mQueuePos >= (int)mQueue.size()) return -1;
    return mQueue[mQueuePos];
}

std::string BenchmarkEngine::statusText() const {
    std::lock_guard<std::mutex> lock(mMutex);
    if (mPhase == Phase::Idle || mPhase == Phase::Done) return "";
    int total = (int)mQueue.size();
    int pos = std::min(mQueuePos + 1, total);
    const std::string& name = (mQueuePos < (int)mQueue.size() && mQueue[mQueuePos] < (int)mNames.size())
                              ? mNames[mQueue[mQueuePos]] : std::string("?");
    const char* phaseName = "";
    double dur = 0;
    switch (mPhase) {
        case Phase::Warmup: phaseName = "warmup"; dur = kWarmupSec; break;
        case Phase::Measure: phaseName = "measure"; dur = kMeasureSec; break;
        case Phase::Cooldown: phaseName = "cooldown"; dur = kCooldownSec; break;
        default: break;
    }
    double elapsed = std::chrono::duration<double>(Clock::now() - mPhaseStart).count();
    double remaining = std::max(0.0, dur - elapsed);
    char buf[160];
    std::snprintf(buf, sizeof(buf), "Bench %d/%d  %s  %s %.1fs",
                  pos, total, name.c_str(), phaseName, remaining);
    return std::string(buf);
}

ShaderResult BenchmarkEngine::makeResultLocked(int idx, const std::vector<double>& times, bool skipped) const {
    ShaderResult r{};
    r.index = idx;
    r.name = idx < (int)mNames.size() ? mNames[idx] : "?";
    int count = (int)times.size();
    if (skipped || count == 0) {
        r.skipped = skipped;
        return r;
    }
    double total = 0;
    for (double t : times) total += t;
    r.avgFps = (double)count / total;
    std::vector<double> sorted = times;
    std::sort(sorted.begin(), sorted.end());
    r.medianFrameMs = sorted[count / 2] * 1000.0;
    int p99idx = std::min(count - 1, (int)(count * 0.99));
    r.p99FrameMs = sorted[p99idx] * 1000.0;
    int lowCount = std::max(1, count / 100);
    double lowSum = 0;
    for (int i = count - lowCount; i < count; ++i) lowSum += sorted[i];
    double lowAvg = lowSum / lowCount;
    r.onePctLowFps = lowAvg > 0 ? 1.0 / lowAvg : 0;
    double dropT = sorted[count / 2] * 2.0;
    int dropped = 0;
    for (double t : times) if (t > dropT) dropped++;
    r.droppedFrames = dropped;
    r.frames = count;
    r.skipped = false;
    return r;
}

double BenchmarkEngine::overallScore(int resW, int resH, bool halfRes) const {
    std::lock_guard<std::mutex> lock(mMutex);
    double sumR = 0;
    int n = 0;
    for (const auto& s : mResults) {
        if (!s.skipped && s.avgFps > 0) {
            sumR += 1.0 / s.avgFps;
            n++;
        }
    }
    if (n == 0) return 0;
    double harmonicMean = (double)n / sumR;
    double pixels = (double)resW * (double)resH * (halfRes ? 0.25 : 1.0);
    constexpr double kReferencePixels = 1920.0 * 1080.0;
    return harmonicMean * 100.0 * pixels / kReferencePixels;
}

std::string BenchmarkEngine::reportJson(int resW, int resH, bool halfRes, bool vsync,
                                        const std::string& gpuName,
                                        const std::string& osVersion,
                                        const std::string& model,
                                        const std::string& thermalStart,
                                        const std::string& thermalEnd,
                                        const std::string& timestamp) const {
    std::lock_guard<std::mutex> lock(mMutex);
    std::ostringstream ss;
    ss.precision(6);
    ss << "{\n";
    ss << "  \"version\": 1,\n";
    ss << "  \"timestamp\": \"" << escapeJson(timestamp) << "\",\n";
    ss << "  \"device\": {\n";
    ss << "    \"os\": \"" << escapeJson(osVersion) << "\",\n";
    ss << "    \"model\": \"" << escapeJson(model) << "\",\n";
    ss << "    \"gpu\": \"" << escapeJson(gpuName) << "\"\n";
    ss << "  },\n";
    ss << "  \"config\": {\n";
    ss << "    \"resolution\": [" << resW << ", " << resH << "],\n";
    ss << "    \"halfRes\": " << (halfRes ? "true" : "false") << ",\n";
    ss << "    \"vsync\": " << (vsync ? "true" : "false") << ",\n";
    ss << "    \"warmupSec\": " << kWarmupSec << ",\n";
    ss << "    \"measureSec\": " << kMeasureSec << ",\n";
    ss << "    \"cooldownSec\": " << kCooldownSec << "\n";
    ss << "  },\n";
    ss << "  \"thermalStateStart\": \"" << escapeJson(thermalStart) << "\",\n";
    ss << "  \"thermalStateEnd\": \"" << escapeJson(thermalEnd) << "\",\n";
    ss << "  \"shaders\": [\n";
    for (size_t i = 0; i < mResults.size(); ++i) {
        const auto& s = mResults[i];
        ss << "    {\n";
        ss << "      \"index\": " << s.index << ",\n";
        ss << "      \"name\": \"" << escapeJson(s.name) << "\",\n";
        ss << "      \"avgFps\": " << s.avgFps << ",\n";
        ss << "      \"onePctLowFps\": " << s.onePctLowFps << ",\n";
        ss << "      \"medianFrameMs\": " << s.medianFrameMs << ",\n";
        ss << "      \"p99FrameMs\": " << s.p99FrameMs << ",\n";
        ss << "      \"frames\": " << s.frames << ",\n";
        ss << "      \"droppedFrames\": " << s.droppedFrames << ",\n";
        ss << "      \"skipped\": " << (s.skipped ? "true" : "false") << "\n";
        ss << "    }";
        if (i + 1 < mResults.size()) ss << ",";
        ss << "\n";
    }
    ss << "  ],\n";
    // overallScore: compute inline to avoid recursive lock.
    // effectivePx = resW × resH × (halfRes ? 0.25 : 1.0)
    // score = harmonicMean(avgFps) × 100 × effectivePx / (1920 × 1080)
    double sumR = 0;
    int n = 0;
    for (const auto& s : mResults) {
        if (!s.skipped && s.avgFps > 0) { sumR += 1.0 / s.avgFps; n++; }
    }
    double score = 0;
    if (n > 0) {
        double harmonicMean = (double)n / sumR;
        double pixels = (double)resW * (double)resH * (halfRes ? 0.25 : 1.0);
        constexpr double kReferencePixels = 1920.0 * 1080.0;
        score = harmonicMean * 100.0 * pixels / kReferencePixels;
    }
    ss << "  \"overallScore\": " << score << "\n";
    ss << "}\n";
    return ss.str();
}

} // namespace bench
